doc/contributing/adding-v8-fast-api.md

```diff
@@ -1,91 +1,396 @@
+
+The following are valid argument types in Fast API callback signatures:
+
+* `bool`
+* `int32_t`
+* `uint32_t`
+* `int64_t`
+* `uint64_t`
+* `float`
+* `double`
+* `v8::Local<v8::Value>` (analogous to `any`)
+* `v8::FastOneByteString&` (analogous to `string`, but _only_ allows sequential
+  one-byte strings, which is often not useful)
+
+<!--
+Deliberately omitted:
+* `void *` (let's not go down this road...)
+* `v8::Local<v8::Object>` (V8 advises that this should only be used for
+                           receiver arguments, but it's more consistent to use
+                           v8::Value handles throughout)
+ -->
+
+The list of valid return types is similar:
+
+* `void`
+* `bool`
+* `int32_t`
+* `uint32_t`
+* `int64_t`
+* `uint64_t`
+* `float`
+* `double`
+
+<!-- * `void *` -->
+
+### Prepending a `receiver` argument
+
+If the first argument of the fast callback signature is of type
+`v8::Local<v8::Value>`, then V8 will always pass the receiver (the `this` value
+of the JavaScript function call) in the first position, meaning that the actual
+arguments will be shifted one position to the right:
+
+```cpp
+// Let's say that this function was bound as a method on some object,
+// such that it would be called in JavaScript as `object.hasProperty(foo)`.
+bool FastHasProperty(v8::Local<v8::Value> receiver,
+                     v8::Local<v8::Value> property,
+                     const v8::FastApiCallbackOptions& options) {
+  v8::Isolate* isolate = options.isolate;
+
+  if (!receiver->IsObject()) {
+    // invalid `this` value; throw some kind of error here
+  }
 
-  ```cpp
-  // Instead of invoking the method as `receiver.internalModuleStat(input)`, the JS land should
-  // invoke it as `internalModuleStat(binding, input)` to make sure the binding is available to
-  // the native land.
-  static int32_t FastInternalModuleStat(
-      Local<Object> unused,
-      Local<Object> recv,
-      const FastOneByteString& input,
-      FastApiCallbackOptions& options) {
-    Environment* env = Environment::GetCurrent(recv->GetCreationContextChecked());
-    // More code
+  bool result;
+  if (!receiver.As<v8::Object>()->Has(isolate->GetCurrentContext(),
+                                      property).To(&result)) {
+    // error pending in V8, value is ignored
+    return false;
   }
-  ```
 
-## Fallback to slow path
+  return result;
+}
+```
+
+The majority of function bindings in the Node.js API do not need to access the
+receiver, as they are intended to be called as a standalone function, not as a
+method on an object or prototype.
+
+However, this still leads to an important caveat if your fast callback takes an
+argument of type `v8::Local<v8::Value>` in the first position. Since V8 will
+always interpret this argument as the receiver (even if the receiver is `null`
+or `undefined`), you will need to ensure that the first argument is discarded:
+
+```cpp
+bool FastIsObject(v8::Local<v8::Value>, // receiver, discarded
```

Comment:
Should the convention be for a named or unnamed discard parameter here? Both are currently used within the codebase.